crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +463,55 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if there is just a single inhabitant for this type.
+    ///
+    /// Note: This function aims to have no false positives, but might return `false`
+    /// for more complicated types that are actually singletons.
+    pub(crate) fn is_singleton(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Any
+            | Type::Never
+            | Type::Unknown
+            | Type::Unbound
+            | Type::Module(..)
+            | Type::Instance(..)
+            | Type::IntLiteral(..)
+            | Type::StringLiteral(..)
+            | Type::BytesLiteral(..)
+            | Type::LiteralString => {
+                // Note: The literal types included in this pattern are not true singletons.
+                // There can be multiple Python objects (at different memory locations) that
+                // are both of type Literal[345], for example.
+                false
+            }
+            Type::None | Type::BooleanLiteral(_) | Type::Function(..) | Type::Class(..) => true,
+            Type::Tuple(tuple) => {
+                !tuple.elements(db).is_empty()
+                    && tuple.elements(db).iter().all(|ty| ty.is_singleton(db))
+            }
+            Type::Union(..) => {
+                // There are some rare edge cases where a union type might be a singleton type.
+                // For example, a union with just one element (which itself is a singleton). Or
+                // a union with an empty type (e.g. Never | None). Here, we assume that such
+                // types would have been simplified to a different representation earlier and
+                // simply return false.
+                false
```

Comment:
I can try to do better here, but I'm not sure if it's worth it.

Reply:
> We normalize unions containing a single type to just that type in:

I saw that, but wasn't sure if there was a guarantee/invariant on the `Type` enum that would enforce this?

> I would have to look into e.g. `Never | None` and @AlexWaygood might know the answer but I would expect this to normalize to `Never`

I think it should normalize to `None`? I'll check and add a test.

Reply:
Yeah, reading through `is_subtype` I think `None` is correct.

Reply:
> I think it should normalize to `None`? I'll check and add a test.

Yes. Every Python type represents a set of possible runtime values. The `None` type is a set of possible values with length 1, and the `Never` type is a set of possible values with length 0. The union of those two sets is exactly equal to the set that represents the `None` type; hence `None | Never` simplifies to `None` (and in fact, `T | Never` always simplifies to `T`, for any given type `T`)

Reply:
I think I would probably shorten the comment here to something like

```suggestion
                // A single-element union, where the sole element was a singleton,
                // would itself be a singleton type. However, unions with length < 2
                // should never appear in our model due to [`UnionBuilder::build`]
                false
```